### PR TITLE
chore(testing): adds enzyme and updates existing tests

### LINKED ISSUE #275: Visual Regression testing
I'd like to open this issue for conversation (since I've heard a few folks start to mention it). Visual Regression tests give us more detailed information (pixel-to-pixel) and granular level details about specific browsers. They also keep us aware of CSS changes happening upstream in PF Core (and let us know when something changes here in PF React that is expected/unexpected). 

There are many tools available, but one I've heard mentioned a few times and would really like to explore further w/ Storybook is [percy.io](https://percy.io/). There is already a storybook [guide](https://percy.io/docs/clients/javascript/react-storybook) so setup would be extremely easy.

I view these tests more as **informational** in PatternFly React (and less critical than they are in PF Core where the design application is happening). From the Percy [documentation notes](https://percy.io/docs/integrations/github):

> Percy provides a workflow for visual review alongside code review. If a build has visual diffs, it will be marked as a failure in the Pull Request. You can review and approve visual diffs in one click as part of your code review process.

> Approving a build is not required, you can still merge the Pull Request in GitHub. Approving a build in Percy simply sets the GitHub commit status to green/successful, so that your team can see that any visual changes were reviewed and approved.

So essentially, we would be notified and could approve or bypass (thus not blocking our dev build/release builds at any point).

@jeff-phillips-18 @Rohoover @jgiardino @LHinson @maryclarke 

I'm curious to hear your thoughts on this...

Also - if the community has any other tools along these lines they'd like to suggest, it would be great to know about them.

### PR DESCRIPTION
**Pull request by AllenBW**
closes https://github.com/patternfly/patternfly-react/issues/257

Yah might be wondering, why doesn't this pr drastically increase test coverage!  Scope. While many test files were touched, the changes are similar and new tests have not been written (ok well there are a few, but not many)

This also updates jest so we can use super fancy snapshotSerializers

This also adds a few needed coverage reporting deps semi related to: #https://github.com/facebook/jest/issues/5772

I'd really like to pull `"react-test-renderer": "^16.2.0"` from 📦 butttttt as new features are merged, they are using the old testing ⚙️, I just can't keep up 😩

### CODE REVIEW

File: src/components/AboutModal/__snapshots__/AboutModal.test.js.snap [hunk 1]
```diff
@@ -1,3 +1,5 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
 exports[`AboutModal renders correctly when hidden 1`] = `null`;
+
+exports[`AboutModal renders correctly when shown 1`] = `"<div role=\\"dialog\\"><div class=\\"modal-backdrop fade in\\"></div><div role=\\"dialog\\" tabindex=\\"-1\\" style=\\"display: block; padding-right: 0px;\\" class=\\"fade in modal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content about-modal-pf\\" role=\\"document\\"><div class=\\"modal-header\\"><button class=\\"close\\"><span aria-hidden=\\"true\\" class=\\"pficon pficon-close\\" title=\\"Close\\"></span><span class=\\"sr-only\\">Close</span></button></div><div class=\\"modal-body\\"><h1>Product Title</h1><div class=\\"product-versions-pf\\"><ul class=\\"list-unstyled\\"><li class=\\"\\"><strong>Label</strong> Version</li></ul></div><div class=\\"trademark-pf\\">Trademark and Copyright Information</div></div><div class=\\"modal-footer\\"><img src=\\"\\" alt=\\"Patternfly Logo\\"></div></div></div></div></div>"`;
```
Comment (priley86) [inline src/components/AboutModal/__snapshots__/AboutModal.test.js.snap]: is there any way to clean up this snapshot? the way `component.html()` is snapping is not so pretty...

**Comment by priley86** [comment]
@AllenBW do you mind summarizing your thoughts thus far? I guess i am still seeing Enzyme as a supplement to jest snapshots (not a replacement)...from the Jest docs: https://facebook.github.io/jest/docs/en/snapshot-testing.html#does-snapshot-testing-substitute-unit-testing). So this helps a lot with the event based/interactive tests, but we can get a lot of coverage from the simple prop based/snapshot based tests. I don't see the need to duplicate tests "just because" of older conventions I guess is what I am after... The snapshots scale really well w/ regressions (as we can update them quickly and visually see what's changing). For example, if you were to change the default class of the `Button` in this project, you would be able to immediately see every impacted component and that component's snapshots. You would also be able to immediately update them with one command. Not that is scale at work! ;)

**Comment by priley86** [comment]
@AllenBW maybe you'd like to present on this topic at a future PF React Community meeting?

File: src/components/AboutModal/__snapshots__/AboutModal.test.js.snap (hunk 1 above)
Comment (AllenBW) [reply src/components/AboutModal/__snapshots__/AboutModal.test.js.snap]: When you say "not so pretty" and "clean up" I understand your meaning to be, that the single line escaped html output is difficult to read at a glance.  The formatting is shite, is that about right? 

Or is your issue with the type of output?  I would argue the content rendered by a component *far* more meaningful than alternatives what is rendered when prop `show = true`.

**Comment by AllenBW** [comment]
@priley86  **tl;dr** Thoughts, greater flexibility, less code, more natural setup and use, works to make snapshots more meaningful.

This is not a replacement for jest snapshots, if for anything,  possibly `react-test-renderer`. This gives us more flexibility with the way we instantiate components (shallow mount or render) and a greater number of ways we can interact with those instances. I have found working with enzyme to be more "natural" and FAR more concise.  Lets say I want to test a component button click...

```
import React from 'react'
import {mount} from 'enzyme'
import MyComponent from './foo'

it('foo should be able to click the button', () => {
    const mockFn = jest.fn();
    const element = mount(<Foo someAction={mockFn} />);
    element.find('button').simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });
```
As yah can see, code is require, reading along is just a lil bit more natural.

Also yah, if this is something that sticks, yeah wouldn't mind presenting pfr community meeting.

Our coverage sucks. Why does our coverage suck? Because functions are not invoked during testing. Why are functions not being invoked during testing?  Because stubbing live components yah can poke n prod can be tricky.  Enzyme, this work, makes it less tricky by being able to either mount render or shallow instantiate components.  With the apis each of those wrappers expose, devs are able to do a BUNCH of things, ones that I have done in this work, (still more to do, gimme a break i'm new at this one) include rendering components as html (rather than a giant jsx-ish json) assess the outcome of props rather than just their values...





🌮 💃 🥑

**Comment by AllenBW** [comment]
ps, still a wip. 
pss, @priley86 thank you for talking 🤗 🙇‍♀️ 🍦

**Comment by priley86** [comment]
I really appreciate all this W.I.P. discussion candy... it makes me so hungry though haha ^^  🍨 🍬 😋

File: src/components/AboutModal/__snapshots__/AboutModal.test.js.snap (hunk 1 above)
Comment (sharvit) [reply src/components/AboutModal/__snapshots__/AboutModal.test.js.snap]: Se how we use `toJson` in foreman:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/Alert/AlertBody.test.js#L11

Comment (AllenBW) [reply src/components/AboutModal/__snapshots__/AboutModal.test.js.snap]: Oh but this isn't the shallow wrapper, its the mount wrapper, already html... we can `JSON.stringify()` to make remove all the escaping... (updated the pr to reflect)

Comment (AllenBW) [reply src/components/AboutModal/__snapshots__/AboutModal.test.js.snap]: @sharvit greaaaaaaaaat suggestion!  while the wrapper type is different, `mount` vs `shallow` used in foreman... it looks like the `.render()` output yields the same result... will go back and fixup a few other instances in this pr on monday

**Comment by sharvit** [comment]
@AllenBW I think you can learn a lot about enzyme/jest/snapshots from https://github.com/theforeman/foreman/pull/5240

Especially from:
https://github.com/theforeman/foreman/pull/5240/files#diff-736af63fce546e2ed852e81792b1358e
https://github.com/theforeman/foreman/pull/5240/files#diff-0a888503d79311bb9407605e191eae95

**Comment by priley86** [comment]
indirectly related... but we still need to find a way to test the react-overlays / Modals as well...and just mentioning an upstream response on this yesterday (not sure that it helps yet):

https://github.com/react-bootstrap/react-overlays/issues/225#issuecomment-371246442

https://github.com/patternfly/patternfly-react/blob/master/src/components/AboutModal/AboutModal.test.js

**Comment by AllenBW** [comment]
@priley86 did yah checkout [the about modal snapshot?](https://github.com/AllenBW/patternfly-react/blob/2e7b768ec2eda5b6e6800cb0d29f3ff9a103e24b/src/components/AboutModal/__snapshots__/AboutModal.test.js.snap#L5) when a component is wrapped in a mount we're able to interact with it throughout the lifecycle

**Comment by priley86** [comment]
i'm just watching the test coverage go up up up before my  👁 👁  's ... this is incredible...

**Comment by priley86** [comment]
@AllenBW please feel free to continue the Enzyme magic as you have time...this was a huge huge win for us already.

**Comment by AllenBW** [comment]
Thats the plan!! 😋 🙇‍♀️

**Pull request merged**